src/include/control_points.h

```diff
@@ -0,0 +1,106 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#pragma once
+
+/* control_points.h: Declarations for control points. */
+
+/* These types are not conditional upon HAVE_CONTROL_POINT since they are used in extern.h. */
+
+#if defined(DOXYGEN) || defined(SWIG)
+#define __F(func) func
+#else
+/* NOLINTNEXTLINE(misc-macro-parentheses) */
+#define __F(func) *func
+#endif
+
+/*!
+ * Identifies a per connection control point or a per session control point.
+ *
+ * A WT_CONTROL_POINT_ID does not specify whether a control point is
+ * a per connection control point or a per session control point since both
+ * start numbering at 0.
+ *
+ * - The maximum per connection control point ID is CONNECTION_CONTROL_POINTS_SIZE - 1.
+ * - The maximum per session control point ID is SESSION_CONTROL_POINTS_SIZE - 1.
+ */
+typedef int32_t wt_control_point_id_t;
+
+/*!
+ * Identifies a control point action.
+ * Used to verify compatibility between the code at a
+ * control point call site and at the control point trigger site.
+ * Zero if the action does not have call site code.
+ */
+typedef uint32_t wt_control_point_action_id_t;
+
+/*!
+ * A function to initialize a control point's data.
+ */
+typedef WT_CONTROL_POINT *wt_control_point_init_t(
+  WT_SESSION_IMPL *session, const char *cp_config_name, const char **cfg);
+
+/*!
+ * A function to test whether a control point should be triggered.
+ */
+typedef bool wt_control_point_pred_t(
+  WT_SESSION_IMPL *session, WT_CONTROL_POINT_REGISTRY *cp_registry, WT_CONTROL_POINT *data);
+
+/*!
+ * Registration data for one control point.
+ */
+struct __wt_control_point_registry {
+    wt_control_point_init_t __F(init); /* Function to initialize the control point. */
+    wt_control_point_pred_t __F(pred); /* Function to test whether to trigger. */
+    size_t crossing_count;             /* Count of executions of the trigger site. */
+    size_t trigger_count;              /* Count of triggers, i.e. pred returned true. */
+    WT_SPINLOCK lock;                  /* Atomically access data and data->ref_count. */
+    const char *config_name;           /* Control point config name */
+    WT_CONTROL_POINT *data; /* Disabled if NULL. More data may follow WT_CONTROL_POINT. */
```

Comment:
I think we could make the code more readible by improving variable names, `data` is not a good variable name. Given this is the control point registry we could name this variable `points`. Then when we read from it instead of:

```
    WT_CONTROL_POINT *saved_data;
    saved_data = cp_registry->data;
```

We could have:
```
    WT_CONTROL_POINT *control_point;
    control_points = cp_registry->points;
```

This would make the code in `cp/` a lot clearer. I was wondering for a while what "data" was but it's just a set of control point objects.

Reply:
Though data isn't exactly a control point either, theres the registry and the data which I think is the point itself is there a way we can make this clearer?

Reply:
The data about a control point is divided into two pieces:
* WT_CONTROL_POINT_REGISTRY which contains data about registering a control point.
* WT_CONTROL_POINT and following which contains the data (configuration parameters and state variables) about a control point.

I suggest the following:
* Rename WT_CONTROL_POINT to WT_CONTROL_POINT_DATA. This is similar to WT_CONTROL_POINT_REGISTRY.
* Once that is done I don't think data is a bad name. However, I suggest using cp_data instead of data.
* control_point is not a good name.  It is too similar to WT_CONNECTION_IMPL.control_points and WT_SESSION_IMPL.control_points.

I will implement my suggestion.

Reply:
Done